File: FastMines_core/UnitTest.FastMines.Core/MosaicMVC/MosaicViewTest.cs
using System;
using System.Threading.Tasks;
using System.ComponentModel;
using System.Reactive.Linq;
using System.Collections.Generic;
using NUnit.Framework;
using fmg.common;
using fmg.common.ui;
using fmg.common.geom;
using fmg.core.mosaic.cells;
using DummyImage = System.Object;
using MosaicTestModel = fmg.core.mosaic.MosaicDrawModel<object>;

namespace fmg.core.mosaic {

    class MosaicTestView : MosaicView<DummyImage, DummyImage, MosaicTestModel> {
        internal MosaicTestView() : base(new MosaicTestModel()) { }
        protected override DummyImage CreateImage() { return new DummyImage(); }
        internal int DrawCount { get; private set; }
        protected override void DrawModified(ICollection<BaseCell> modifiedCells) {
            LoggerSimple.Put(nameof(MosaicTestView) + "::DrawModified");
            ++DrawCount;
        }
        protected override void Disposing() {
            base.Disposing();
            Model.Dispose();
        }
    }

    public class MosaicViewTest {

        internal const int TEST_SIZE_W = MosaicModelTest.TEST_SIZE_W;
        internal const int TEST_SIZE_H = MosaicModelTest.TEST_SIZE_H;

        [SetUp]
        public void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put("> MosaicViewTest::Setup");

            MosaicModelTest.StaticInitializer();

          //Observable.Just("UI factory inited...").Subscribe(LoggerSimple.Put);
        }

        [SetUp]
        public void Before() {
            LoggerSimple.Put("======================================================");
        }

        [OneTimeTearDown]
        public void After() {
            LoggerSimple.Put("======================================================");
            LoggerSimple.Put("< " + nameof(MosaicViewTest) + " closed");
            LoggerSimple.Put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        [Test]
        public async Task PropertyChangedTest() {
            var modifiedProperties = new List<string>();
            void onViewPropertyChanged(object sender, PropertyChangedEventArgs ev) {
                LoggerSimple.Put("  MosaicTestView::PropertyChangedTest: onViewPropertyChanged: ev.name=" + ev.PropertyName);
                modifiedProperties.Add(ev.PropertyName);
            }
            using (var view = new MosaicTestView()) {
                view.PropertyChanged += onViewPropertyChanged;

                Factory.DEFERR_INVOKER(() => view.Model.Size = new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));

                await Task.Delay(200);

                Assert.IsTrue(modifiedProperties.Contains(nameof(view.Model)));
                Assert.IsTrue(modifiedProperties.Contains(nameof(view.Size)));
                Assert.IsTrue(modifiedProperties.Contains(nameof(view.Image)));

                view.PropertyChanged -= onViewPropertyChanged;
            }
        }

        [Test]
        public void ReadinessAtTheStartTest() {
            using (var view = new MosaicTestView()) {
                Assert.AreEqual(0, view.DrawCount);
                Assert.NotNull(view.Image);
                Assert.AreEqual(1, view.DrawCount);
            }
        }

        [Test]
        public async Task MultipleChangeModelOneDrawViewTest() {
            using (var view = new MosaicTestView()) {
                Assert.AreEqual(0, view.DrawCount);

                var m = view.Model;
                Factory.DEFERR_INVOKER(() => MosaicModelTest.ChangeModel(m));
                await Task.Delay(100);

                var img = view.Image;
                Assert.NotNull(img);
                Assert.AreEqual(1, view.DrawCount);

                // test no change
                Factory.DEFERR_INVOKER(() => m.Size = new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));
                await Task.Delay(100);
                Assert.AreEqual(true, ReferenceEquals(img, view.Image));
                Assert.AreEqual(1, view.DrawCount);

                // test change
                Factory.DEFERR_INVOKER(() => m.Size = new SizeDouble(TEST_SIZE_W + 1, TEST_SIZE_H));
                await Task.Delay(200);
                Assert.AreNotEqual(img, view.Image);
                Assert.NotNull(view.Image);
                Assert.AreEqual(2, view.DrawCount);
            }
        }

        [Test]
        public async Task OneNotificationOfImageChangedTest() {
            using (var view = new MosaicTestView()) {
                var modifiedProperties = new Dictionary<string /* property name */, int /* count */>();

                using (var signal = new Signal()) {
                    var ob = Observable
                        .FromEventPattern<PropertyChangedEventHandler, PropertyChangedEventArgs>(h => view.PropertyChanged += h, h => view.PropertyChanged -= h);
                    using (ob.Subscribe(
                        ev => {
                            var name = ev.EventArgs.PropertyName;
                            LoggerSimple.Put("  OneNotificationOfImageChangedTest: onViewPropertyChanged: ev.name=" + name);
                            modifiedProperties[name] = 1 + (modifiedProperties.ContainsKey(name) ? modifiedProperties[name] : 0);
                        }))
                    {
                        using (ob.Timeout(TimeSpan.FromMilliseconds(50))
                            .Subscribe(ev => {
                                LoggerSimple.Put("onNext:");
                            }, ex => {
                                LoggerSimple.Put("onError: " + ex);
                                signal.Set();
                            }))
                        {
                            Factory.DEFERR_INVOKER(() => MosaicModelTest.ChangeModel(view.Model));

                            Assert.IsTrue(await signal.Wait(TimeSpan.FromSeconds(1)));
                        }
                    }
                }

                LoggerSimple.Put("  OneNotificationOfImageChangedTest: checking...");
                Assert.AreEqual(1, modifiedProperties[nameof(view.Image)]);
                Assert.AreEqual(0, view.DrawCount);
                var img = view.Image; // call the implicit draw method
                Assert.AreEqual(1, view.DrawCount);
            }
        }

    }

}

File: FastMines_core/src_test/fmg/core/mosaic/MosaicViewTest.java
package fmg.core.mosaic;

import java.util.Collection;

import org.junit.*;

import fmg.common.Logger;
import fmg.common.geom.SizeDouble;
import fmg.common.notifier.PropertyChangeExecutor;
import fmg.core.img.IImageView;
import fmg.core.mosaic.cells.BaseCell;
import io.reactivex.Flowable;

class MosaicTestView extends MosaicView<DummyImage, DummyImage, MosaicTestModel> {
    MosaicTestView() { super(new MosaicTestModel()); }
    @Override protected DummyImage createImage() { return new DummyImage(); }
    private int drawCount;
    int getDrawCount() { return drawCount; }
    @Override protected void drawModified(Collection<BaseCell> modifiedCells) {
        Logger.info("MosaicTestView::drawModified");
        ++drawCount;
    }
    @Override public void close() {
        super.close();
        getModel().close();
    }
}

public class MosaicViewTest {

    static final int TEST_SIZE_W = MosaicModelTest.TEST_SIZE_W;
    static final int TEST_SIZE_H = MosaicModelTest.TEST_SIZE_H;

    @BeforeClass
    public static void setup() {
        Logger.info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        Logger.info("> MosaicViewTest::setup");

        MosaicModelTest.ProjSettings();

        Flowable.just("UI factory inited...").subscribe(Logger::info);
    }

    @Before
    public void before() {
        Logger.info("======================================================");
    }
    @AfterClass
    public static void after() {
        Logger.info("======================================================");
        Logger.info("< MosaicViewTest closed");
        Logger.info("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }

    @Test
    public void propertyChangedTest() {
        Logger.info("> MosaicTestView::propertyChangedTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(100, 1000,
            view -> {
                view.getModel().setSize(new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));
            }, (view, modifiedProperties) -> {
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_MODEL));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_MODEL).first);
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_SIZE ));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_SIZE ).first);
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_IMAGE));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_IMAGE).first);
                Assert.assertEquals(                3 , modifiedProperties.size());
            });
    }

    @Test
    public void readinessAtTheStartTest() {
        Logger.info("> MosaicTestView::readinessAtTheStartTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(10, 1000,
            view -> {
                Assert.assertEquals(0, view.getDrawCount());
                Assert.assertNotNull(view.getImage());
                Assert.assertEquals(1, view.getDrawCount());
            }, (view, modifiedProperties) -> { });
    }

    @Test
    public void multipleChangeModelOneDrawViewTest() throws InterruptedException {
        Logger.info("> MosaicTestView::multipleChangeModelOneDrawViewTest");

        DummyImage[] img = { null };
        MosaicTestView[] v = { null };
        new PropertyChangeExecutor<>(() -> v[0] = new MosaicTestView(), false).run(100, 1000,
           view -> {
               Assert.assertEquals(0, view.getDrawCount());
               MosaicModelTest.changeModel(view.getModel());
           }, (view, modifiedProperties) -> {
               img[0] = view.getImage();
               Assert.assertNotNull(img[0]);
               Assert.assertEquals(1, view.getDrawCount());
           });

        // test no change
        new PropertyChangeExecutor<>(() -> v[0], false).run(100, 1000,
           view -> {
               view.setSize(new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));
           }, (view, modifiedProperties) -> {
               Assert.assertEquals(img[0], view.getImage());
               Assert.assertEquals(1, view.getDrawCount());
           });

        // test change
        new PropertyChangeExecutor<>(() -> v[0]).run(100, 1000,
           view -> {
               view.setSize(new SizeDouble(TEST_SIZE_W + 1, TEST_SIZE_H));
           }, (view, modifiedProperties) -> {
               Assert.assertNotEquals(img[0], view.getImage());
               Assert.assertNotNull(view.getImage());
               Assert.assertEquals(2, view.getDrawCount());
           });
    }

    @Test
    public void oneNotificationOfImageChangedTest() {
        Logger.info("> MosaicTestView::oneNotificationOfImageChangedTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(100, 1000,
           view -> {
               MosaicModelTest.changeModel(view.getModel());
           }, (view, modifiedProperties) -> {
               Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_IMAGE));
               Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_IMAGE).first);
               Assert.assertEquals(0, view.getDrawCount());
               view.getImage(); // call the implicit draw method
               Assert.assertEquals(1, view.getDrawCount());
           });
    }

}
